File: src/falloff.js
// Falloff math for floats: where an object LIVES on the days axis (its home,
// the range its own connections span) and how strongly it should register when
// viewed from some distance away from that home. Pure over
// `{document, engine}` and exact throughout -- the lens layer consumes this;
// nothing here touches projection, rendering, or the session.

import { Rational, ZERO } from "./exact.js";
import { effectiveObjectStaples, frameEndDays } from "./staples.js";

/**
 * The exact-Rational day range spanned by an object's staples' resolved frame
 * coordinates -- its implicit placement staple included, because that IS a
 * staple under the connection model (src/staples.js's effectiveObjectStaples).
 * A single resolvable staple gives a zero-width home; a many-valued or
 * object-far end contributes nothing (its instant is not one coordinate);
 * an object with nothing resolvable has no home at all, and `null` is that
 * honest answer -- never a fabricated date.
 *
 * Returns `{startDays, endDays}` (both Rational, startDays <= endDays) or null.
 */
export function objectHome(chronologDocument, engine, objectId) {
  let start = null;
  let end = null;
  for (const row of effectiveObjectStaples(chronologDocument, objectId, engine)) {
    const days = frameEndDays(engine, row.far);
    if (days === null) continue;
    if (start === null || days.compare(start) < 0) start = days;
    if (end === null || days.compare(end) > 0) end = days;
  }
  return start === null ? null : { startDays: start, endDays: end };
}

/**
 * Exact distance in days from a home range to an instant: zero anywhere inside
 * the range (an object is at full presence throughout its own home), the gap
 * to the nearer edge outside it. A null home has no distance to report.
 */
export function distanceFromHome(home, days) {
  if (!home) return null;
  const at = Rational.parse(days);
  if (at.compare(home.startDays) < 0) return home.startDays.sub(at);
  if (at.compare(home.endDays) > 0) return at.sub(home.endDays);
  return ZERO;
}

// The one falloff parameter: HALF_DISTANCE_DAYS days from home, apparent
// magnitude is half the base weight. The curve is the simple smooth reciprocal
//
//   apparent = base * 1 / (1 + distance / halfDistanceDays)
//
// -- monotone decreasing in distance, exactly `base` at distance zero, never
// reaching zero (a float lapses from prominence, it never lapses from truth).
// Chosen over an exponential half-life for its exactness: every value is one
// rational multiply, no transcendental approximation to round.
export const DEFAULT_HALF_DISTANCE_DAYS = "7";

/**
 * The apparent magnitude of `baseWeight` seen from `distanceDays` away.
 *
 * All three inputs parse through Rational (numbers, strings, Rationals), the
 * result IS a Rational -- callers that feed the display-weight machinery call
 * `.toNumber()` at their own boundary. Negative distance is read as its
 * magnitude (distance is a gap, not a direction), and a non-positive
 * `halfDistanceDays` is refused: a falloff that halves at zero distance is a
 * division by zero wearing a parameter.
 */
export function apparentMagnitude(baseWeight, distanceDays, { halfDistanceDays = DEFAULT_HALF_DISTANCE_DAYS } = {}) {
  const base = Rational.parse(baseWeight);
  const distance = Rational.parse(distanceDays).abs();
  const half = Rational.parse(halfDistanceDays);
  if (half.compare(0) <= 0) throw new RangeError("halfDistanceDays must be positive");
  return base.mul(half).div(half.add(distance));
}

File: app/lib/core/falloff.dart
// Pure Dart. Nothing under lib/core/ may import Flutter or a third-party
// package.
//
// Where an object LIVES on the days axis (its home -- the range its own
// connections span) and how strongly it registers when viewed from some
// distance away. Exact throughout: every value here is one rational multiply,
// with no transcendental to approximate and nothing to round.
//
// Ruled generalization (JC-8, Don 2026-08-27, ruling 10): falloff is not a
// special case wired into one lens. It is the CLOSING STEP of the weight
// chain -- "apparent-magnitude falloff as the projector's own closing step,"
// applied multiplicatively after the frame-formula fold. `weight.dart`'s
// `composeWeight` is where that happens; this file is the math it calls.

import 'exact.dart';

/// A closed range on the shared days axis, `start <= end`.
typedef DayExtent = ({Rational start, Rational end});

/// Resolves one staple far end to the day range it occupies, or null when it
/// does not resolve to a coordinate at all.
///
/// THE SEAM: staples are not this file's business. The caller -- the staple
/// substrate, which owns `effectiveObjectStaples` and frame-coordinate
/// resolution -- hands in the far-end ids and this resolver. A many-valued
/// far end, or one that lands on another object rather than a coordinate,
/// resolves to null: its instant is not one coordinate, so it contributes
/// nothing rather than a fabricated one.
typedef ExtentResolver = DayExtent? Function(String farEndId);

/// The day range an object's resolvable staples span -- its implicit
/// placement staple included, because that IS a staple under the connection
/// model. A single resolvable staple gives a zero-width home; an object with
/// nothing resolvable has NO home, and null is that honest answer rather than
/// a fabricated date.
DayExtent? objectHome(Iterable<String> farEndIds, ExtentResolver resolve) {
  Rational? start;
  Rational? end;
  for (final id in farEndIds) {
    final extent = resolve(id);
    if (extent == null) continue;
    if (start == null || extent.start < start) start = extent.start;
    if (end == null || extent.end > end) end = extent.end;
  }
  return start == null ? null : (start: start, end: end!);
}

/// Exact distance in days from a home to an instant: zero anywhere inside the
/// range -- an object is at full presence throughout its own home -- and the
/// gap to the nearer edge outside it. A null home has no distance to report.
Rational? distanceFromHome(DayExtent? home, Rational days) {
  if (home == null) return null;
  if (days < home.start) return home.start - days;
  if (days > home.end) return days - home.end;
  return Rational.zero;
}

/// The one falloff parameter: this many days from home, apparent magnitude is
/// half the base. Seven days is a default, not a law -- a projector may name
/// its own.
final Rational defaultHalfDistanceDays = Rational.fromInt(7);

/// The apparent magnitude of [base] seen from [distance] away:
///
///   apparent = base * half / (half + distance)
///
/// Monotone decreasing in distance, exactly [base] at distance zero, and for
/// a positive base never reaching zero -- a float lapses from prominence, it
/// never lapses from truth. Negative distance reads as its magnitude
/// (distance is a gap, not a direction), and a non-positive half-distance is
/// refused: a falloff that halves at zero distance is a division by zero
/// wearing a parameter.
Rational apparentMagnitude(Rational base, Rational distance, {Rational? halfDistance}) {
  final half = halfDistance ?? defaultHalfDistanceDays;
  if (half <= Rational.zero) {
    throw RangeError('Falloff half-distance must be positive');
  }
  return base * half / (half + distance.abs());
}
